### PR TITLE
bpo-43224: Implement substitution of unpacked TypeVarTuple in C

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue43224

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
Thanks for working on this! Please don't merge this without approval from the PEP 646 authors or the typing experts; I'm not confident we'll get all the edge cases right. I'd have liked a chance to review #31800 in more detail too.

File: Objects/genericaliasobject.c [hunk 1]
```diff
@@ -283,11 +331,27 @@ _Py_subs_parameters(PyObject *self, PyObject *args, PyObject *parameters, PyObje
...
+    for (; varparam < nparams; varparam++) {
+        PyObject *param = PyTuple_GET_ITEM(parameters, varparam);
+        if (Py_TYPE(param)->tp_iter) {  // TypeVarTuple
+            break;
+        }
+    }
+    if (varparam < nparams) {
+        if (nitems < nparams - 1) {
+            return PyErr_Format(PyExc_TypeError,
+                                "!Too few arguments for %R",
```
Comment (JelleZijlstra) [inline Objects/genericaliasobject.c]:
```suggestion
                                "Too few arguments for %R",
```
Or does `!` do something here?

File: Lib/test/test_typing.py [hunk 2]
```diff
@@ -467,7 +467,7 @@ def test_var_substitution(self):
         T2 = TypeVar('T2')
         class G(Generic[Unpack[Ts]]): pass
 
-        for A in G, Tuple:
+        for A in G, Tuple, tuple:
```
Comment (JelleZijlstra) [inline Lib/test/test_typing.py:948]: There need to be more tests for cases where we substitute an unpacked tuple into the TypeVarTuple. I found a crash on your branch:

```
 % ./python.exe 
Python 3.11.0a6+ (heads/serhiy-storchaka-typevartuple-subst-c:a35af462a3, Mar 11 2022, 19:02:45) [Clang 13.0.0 (clang-1300.0.29.3)] on darwin
Type "help", "copyright", "credits" or "license" for more information
>>> from typing import Unpack, TypeVarTuple
>>> Ts = TypeVarTuple("Ts")
>>> t = tuple[Ts]
>>> t[int, Unpack[tuple[int, str]], str]
Assertion failed: (iparam != varparam), function _Py_subs_parameters, file genericaliasobject.c, line 382.
zsh: abort      ./python.exe
```

**Review by JelleZijlstra — CHANGES_REQUESTED** [review]
I found a crash; see comment.

In general I'm worried that this approach will be too difficult to get right in all cases. I'd prefer the approach in #31804, which I suggested in https://github.com/python/cpython/pull/31021#pullrequestreview-890627941.

File: Lib/test/test_typing.py (hunk 2 above)
Comment (serhiy-storchaka) [reply Lib/test/test_typing.py:948]: Good catch. Is not `tuple[Ts]` illegal?

I have added a runtime error for this case. It is consistent for Tuple, tuple and user generics.

File: Objects/genericaliasobject.c (hunk 1 above)
Comment (serhiy-storchaka) [reply Objects/genericaliasobject.c]: It was temporary, for debugging.

**Comment by serhiy-storchaka** [comment]
I have made the requested changes; please review again.

**Comment by mrahtz** [comment]
Wow, thanks for implementing this - it hadn't even occurred to me this would be needed!

I also feel uneasy, though - as I mentioned in https://github.com/python/cpython/pull/31800 we'd decided against implementing it this way. I think we should pause work on this until we've decided on exactly which approach we're proceeding with.

**Comment by mrahtz** [comment]
Based on your note in https://github.com/python/cpython/pull/31800, I'm guessing you're going to proceed with this anyway. In that case, I think it would be worth adding some tests to this PR along the lines of https://github.com/python/cpython/commit/b9b1c80913f1fbc21ea60257bed6562e18c49802#diff-04d29c98076c2d6bb75921ea9becb26a862544d39b71db87b6e354c759b9305dL794.

**Comment by serhiy-storchaka** [comment]
I do not understand what tests are in question. The change you referred removes tests, not adds new tests.

**Comment by mrahtz** [comment]
Right - I'm proposing adding them _back_, with the changes necessary to also test the code in this PR.

**Review by JelleZijlstra — APPROVED** [review]

**Review by JelleZijlstra — APPROVED** [review]

**Pull request merged**